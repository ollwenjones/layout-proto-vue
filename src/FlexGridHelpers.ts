import { FlexGridCellConfig, FlexGridConfig } from "./store";

const LAYOUT_CLASS_NAME = "dp-flex-layout";
const CONTENT_CELL_CLASS_NAME = "dp-flex-layout__content-cell";
const PREVIEW_CELL_CLASS_NAME = "dp-flex-layout__preview-cell";

const GRID_DESKTOP_BREAKPOINT = 840;
const GRID_TABLET_BREAKPOINT = 520;

export function getDesktopClass(cols: number) {
  return `dp-flex-layout--${cols}-col`;
}

export function getTabletClass(cols: number) {
  return `dp-flex-layout--${cols}-col-tablet`;
}

export function getPhoneClass(cols: number) {
  return `dp-flex-layout--${cols}-col-phone`;
}

export function getFlexCellClasses(cell: FlexGridCellConfig) {
  return [
    getDesktopClass(cell.columns.desktop),
    getTabletClass(cell.columns.tablet),
    getPhoneClass(cell.columns.phone)
  ].join(" ");
}

/**
 * Create the CSS markup for the model.
 * 
 * This also illustrates the flexibility & testability of functional programming.
 * 
 * @param rowGap between rows
 * @param columnGap between columns
 * @param desktopColumns how many colums in desktop break-point
 * @param tabletColumns how many columns in tablet break-point
 * @param phoneColumns how many columns in phone break-point
 */
export function generateStyles(
  rowGap: number,
  columnGap: number,
  desktopColumns: number,
  tabletColumns: number,
  phoneColumns: number
) {
  let styleStrings = [];

  // set cell margins on general class names, for gutters:
  styleStrings.push(`
  .${CONTENT_CELL_CLASS_NAME},
  .${PREVIEW_CELL_CLASS_NAME} {
    margin: ${rowGap / 2}px ${columnGap / 2}px;
  }
  `);

  // add desktop rules
  styleStrings.push(
    `@media (min-width: ${GRID_DESKTOP_BREAKPOINT}px) {`,
    ...createStyleNamesForBreakPoint(desktopColumns, columnGap),
    `}

  `
  );

  // add tablet rules
  styleStrings.push(
    `@media (min-width: ${GRID_TABLET_BREAKPOINT}px) and (max-width: ${GRID_DESKTOP_BREAKPOINT -
      1}px) {`,
    ...createStyleNamesForBreakPoint(tabletColumns, columnGap, "tablet"),
    `}

  `
  );

  // add phone rules
  styleStrings.push(
    `@media (max-width: ${GRID_TABLET_BREAKPOINT - 1}px) {`,
    ...createStyleNamesForBreakPoint(phoneColumns, columnGap, "phone"),
    `}

  `
  );

  return styleStrings.join("");
}

function createStyleNamesForBreakPoint(columns: number, columnGap: number, sizeName = "") {
  const names = [];
  for (let i = columns; i > 0; i--) {
    names.push(`
      .${getClassNameForCol(i, sizeName)} {
        width: calc(${(i / columns) * 100}% - ${columnGap}px);
        /* flex: ${i}; */
      }
    `);
  }
  return names;
}

function getClassNameForCol(spanning: number , sizeName = "") {
  return sizeName
    ? `${LAYOUT_CLASS_NAME}--${spanning}-col-${sizeName}`
    : `${LAYOUT_CLASS_NAME}--${spanning}-col`;
}
