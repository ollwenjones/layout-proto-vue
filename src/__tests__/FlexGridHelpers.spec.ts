import { generateStyles } from "../FlexGridHelpers";

test('generateStyles', () => {
    // since we know via browser testing, that it currently does what we want
    // we can "snapshot" the result to ensure any modifications in the future
    // don't break what's expected here. Not as thorough as a dozen, difficult
    // to write comparisons, but it's very fast to write.
    // If I'm suspicious of this, I can change the params and inspect the output.
    expect(generateStyles(10, 10, 12, 8, 4)).toMatchSnapshot();
});