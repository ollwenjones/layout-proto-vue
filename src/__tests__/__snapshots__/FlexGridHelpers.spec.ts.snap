// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateStyles 1`] = `
"
  .dp-flex-layout__content-cell,
  .dp-flex-layout__preview-cell {
    margin: 5px 5px;
  }
  @media (min-width: 840px) {
      .dp-flex-layout--12-col {
        width: calc(100% - 10px);
        /* flex: 12; */
      }
    
      .dp-flex-layout--11-col {
        width: calc(91.66666666666666% - 10px);
        /* flex: 11; */
      }
    
      .dp-flex-layout--10-col {
        width: calc(83.33333333333334% - 10px);
        /* flex: 10; */
      }
    
      .dp-flex-layout--9-col {
        width: calc(75% - 10px);
        /* flex: 9; */
      }
    
      .dp-flex-layout--8-col {
        width: calc(66.66666666666666% - 10px);
        /* flex: 8; */
      }
    
      .dp-flex-layout--7-col {
        width: calc(58.333333333333336% - 10px);
        /* flex: 7; */
      }
    
      .dp-flex-layout--6-col {
        width: calc(50% - 10px);
        /* flex: 6; */
      }
    
      .dp-flex-layout--5-col {
        width: calc(41.66666666666667% - 10px);
        /* flex: 5; */
      }
    
      .dp-flex-layout--4-col {
        width: calc(33.33333333333333% - 10px);
        /* flex: 4; */
      }
    
      .dp-flex-layout--3-col {
        width: calc(25% - 10px);
        /* flex: 3; */
      }
    
      .dp-flex-layout--2-col {
        width: calc(16.666666666666664% - 10px);
        /* flex: 2; */
      }
    
      .dp-flex-layout--1-col {
        width: calc(8.333333333333332% - 10px);
        /* flex: 1; */
      }
    }

  @media (min-width: 520px) and (max-width: 839px) {
      .dp-flex-layout--8-col-tablet {
        width: calc(100% - 10px);
        /* flex: 8; */
      }
    
      .dp-flex-layout--7-col-tablet {
        width: calc(87.5% - 10px);
        /* flex: 7; */
      }
    
      .dp-flex-layout--6-col-tablet {
        width: calc(75% - 10px);
        /* flex: 6; */
      }
    
      .dp-flex-layout--5-col-tablet {
        width: calc(62.5% - 10px);
        /* flex: 5; */
      }
    
      .dp-flex-layout--4-col-tablet {
        width: calc(50% - 10px);
        /* flex: 4; */
      }
    
      .dp-flex-layout--3-col-tablet {
        width: calc(37.5% - 10px);
        /* flex: 3; */
      }
    
      .dp-flex-layout--2-col-tablet {
        width: calc(25% - 10px);
        /* flex: 2; */
      }
    
      .dp-flex-layout--1-col-tablet {
        width: calc(12.5% - 10px);
        /* flex: 1; */
      }
    }

  @media (max-width: 519px) {
      .dp-flex-layout--4-col-phone {
        width: calc(100% - 10px);
        /* flex: 4; */
      }
    
      .dp-flex-layout--3-col-phone {
        width: calc(75% - 10px);
        /* flex: 3; */
      }
    
      .dp-flex-layout--2-col-phone {
        width: calc(50% - 10px);
        /* flex: 2; */
      }
    
      .dp-flex-layout--1-col-phone {
        width: calc(25% - 10px);
        /* flex: 1; */
      }
    }

  "
`;
